mingpt/mingptmlx.py

```diff
@@ -0,0 +1,309 @@
+
+        return logits, loss
+
+    def generate(self, idx, max_new_tokens):
+        # idx is (B, T) array of indices in the current context
+        for _ in range(max_new_tokens):
+            # crop idx to the last block_size tokens
+            idx_cond = idx[:, -block_size:]
+            # get the predictions
+            logits, _ = self(idx_cond)
+            # focus only on the last time step
+            logits = logits[:, -1, :] # becomes (B, C)
+            # apply softmax to get probabilities
+            probs = nn.softmax(logits, axis=-1) # (B, C)
+            # sample from the distribution
+            idx_next = mx.random.categorical(probs, axis=-1, num_samples=1)
+            # append sampled index to the running sequence
+            idx = mx.concatenate((idx, idx_next), axis=1) # (B, T+1)
+        return idx
+
+
+# %%
+
+model = BigramLanguageModel()
+
+# %%
+params = model.parameters()
+
+
+# %%
+# print the number of parameters in the model
+p = sum(v.size for _, v in tree_flatten(model.parameters())) / 10**6
+print(f"Total parameters {p:.3f}M")
+
+# %%
+# create an optimizer
+optimizer = optim.AdamW(learning_rate)
+
+# %%
+def estimate_loss(model):
+    out = {}
+    for split in ['train', 'val']:
+        losses = mx.zeros(eval_iters)
+        for k in range(eval_iters):
+            X, Y = get_batch(split)
+            logits, loss = model(X, Y)
+            losses[k] = loss.item()
+        out[split] = losses.mean()
+    model.train()
+    return out
+
+def loss_fn(model, x, y):
+    _, loss = model(x, y)
+    return loss
+
+def step(model, optimizer, inputs, targets):
+    loss_and_grad_fn = nn.value_and_grad(model, loss_fn)
+    loss, grads = loss_and_grad_fn(model, inputs, targets)
+    optimizer.update(model, grads)
+    return loss
+
+
+# %%
+model.train()
+for iter in range(max_iters):
+    # every once in a while evaluate the loss on train and val sets
+    if iter % eval_interval == 0 or iter == max_iters - 1:
+        losses = estimate_loss(model)
+        print(f"step {iter}: train loss {losses['train'].item():.4f}, val loss {losses['val'].item():.4f}")
+
+    # sample a batch of data
+    xb, yb = get_batch('train')
+
+    # evaluate the loss
+    step(model, optimizer, xb, yb)
+
+print("Training complete")
+
+# %%
+# save weights
+model.save_weights("./shakespearebigramweights.npz")
+
+# %%
+def generate(model, idx, max_new_tokens):
+    # idx is (B, T) array of indices in the current context
+    for _ in range(max_new_tokens):
+        # crop idx to the last block_size tokens
+        idx_cond = idx[:, -block_size:]
+        # get the predictions
+        logits, _ = model(idx_cond)
+        # focus only on the last time step
+        logits = logits[:, -1, :] # becomes (B, C)
+        # apply softmax to get probabilities
+        probs = nn.softmax(logits, axis=-1) # (B, C)
+        # sample from the distribution
+        idx_next = mx.random.categorical(probs, axis=-1, num_samples=10)
```

Comment:
`mx.categorical` expects unnormalized output. Try removing the `softmax` and simply using the `logits` directly.

Reply:
Thanks a ton Awni! That was the bug in my code.  It looks like the model did train properly and is generating tokens as expected. 

❤️